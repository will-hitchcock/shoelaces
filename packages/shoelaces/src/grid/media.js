import breakpoints, { keys } from '../const/breakpoints'
import deepmerge from 'deepmerge'

const queries = {
  min: key => `@media (min-width: ${breakpoints[key]})`,
  max: key => `@media (max-width: ${breakpoints[key]})`,
  between: (start, end) =>
    `@media (min-width: ${breakpoints[start]} and (max-width: ${breakpoints[
      end
    ]}`
}

export const min = key => args => {
  if (!Object.keys(args).length) return null
  if (!key || !breakpoints[key]) {
    return args
  }
  return { [queries.min(key)]: args }
}

export const max = key => args => {
  if (!Object.keys(args).length) return null
  if (!key) {
    return args
  }
  return { [queries.max(key)]: args }
}

export const between = (start, end) => args => {
  if (!Object.keys(args).length) return null
  if (!start || !end) {
    return args
  }
  return { [queries.between(start, end)]: args }
}

export const doQuery = (query, obj, func) => {
  let processedObj = {}

  if (typeof obj !== 'object') {
    return { ...processedObj, ...func(obj) }
  }

  for (let key in obj) {
    processedObj = { ...processedObj, ...query(key)(func(obj[key])) }
  }

  return processedObj
}

export const deepQuery = (query, arr) =>
  deepmerge.all(arr.map(item => doQuery(query, item.data, item.mixin)))

export default { min, max, between, doQuery, deepQuery }
