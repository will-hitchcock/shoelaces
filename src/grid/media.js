import breakpoints from '../const/breakpoints'

const queries = {
  min: key => `@media (min-width: ${breakpoints[key]})`,
  max: key => `@media (max-width: ${breakpoints[key]})`,
  between: (start, end) =>
    `@media (min-width: ${
      breakpoints[start]
    } and (max-width: ${
      breakpoints[end]
    }`
}

export const min = key => args => {
  if (!key || !breakpoints[key]) { return args }
  return { [queries.min(key)]: args }
}

export const max = key => args => {
  if (!key) { return args }
  return { [queries.max(key)]: args }
}

export const between = (start, end) => args => {
  if (!start || !end) { return args }
  return { [queries.between(start, end)]: args }
}

export const flatten = (query, prop, mixin) => {
  if (typeof prop !== 'object') { return mixin(prop) }
  const output = {}

  Object.keys(breakpoints)
    .filter(a => !!prop[a])
    .map(value => (output[query(value)] = mixin(prop[value])))

  return output
}

export const flattenMin = (...args) => flatten(queries.min, ...args)
export const flattenMax = (...args) => flatten(queries.max, ...args)

export default { min, max, between, flattenMin, flattenMax }
