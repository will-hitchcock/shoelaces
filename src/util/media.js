import breakpoints from '../const/breakpoints'

export const queries = {
  min: key => `@media (min-width: ${breakpoints[key]})`,
  max: key => `@media (max-width: ${breakpoints[key]})`,
  between: (start, end) =>
    `@media (min-width: ${
      breakpoints[start]
    } and (max-width: ${
      breakpoints[end]
    }`
}

export const min = key => args => {
  if (!key) return args
  return { [queries.min(key)]: args }
}

export const max = key => args => {
  if (!key) return args
  return { [queries.max(key)]: args }
}

export const between = (start, end) => args => ({
  [queries.between(start, end)]: args
})

export const flatten = (query, prop, mixin) => {
  if (typeof prop !== 'object') { return mixin(prop) }
  const output = {}

  Object.keys(breakpoints)
    .filter(a => !!prop[a])
    .map(value => (output[query(value)] = mixin(prop[value])))

  return output
}

export const flattenMin = (...args) => flatten(queries.min, ...args)
export const flattenMax = (...args) => flatten(queries.max, ...args)

export default { queries, min, max, between, flatten, flattenMin, flattenMax }
