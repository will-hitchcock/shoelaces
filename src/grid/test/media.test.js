import { min, max, between, flattenMin, flattenMax } from '../media'

describe('media', () => {
  describe('min', () => {
    it('renders the styles without a media query', () => {
      expect(min()({ key: 'value' })).toMatchSnapshot()
    })

    it('renders the styles without a media query', () => {
      expect(min('xs')({ key: 'value' })).toMatchSnapshot()
    })

    it('renders the styles inside a min md media query', () => {
      expect(min('md')({ key: 'value' })).toMatchSnapshot()
    })
  })

  describe('max', () => {
    it('renders the styles without a media query', () => {
      expect(max()({ key: 'value' })).toMatchSnapshot()
    })

    it('renders the styles inside a max md media query', () => {
      expect(max('md')({ key: 'value' })).toMatchSnapshot()
    })
  })

  describe('between', () => {
    it('renders the styles without a media query', () => {
      expect(between()({ key: 'value' })).toMatchSnapshot()
    })

    it('renders the styles without a media query', () => {
      expect(between('xs')({ key: 'value' })).toMatchSnapshot()
    })

    it('renders the styles without a media query', () => {
      expect(between(null, 'lg')({ key: 'value' })).toMatchSnapshot()
    })

    it('renders the styles inside a between md and lg media query', () => {
      expect(between('md', 'lg')({ key: 'value' })).toMatchSnapshot()
    })
  })

  describe('flattenMin', () => {
    it('renders the renders width 0.5em inside a min xs media query', () => {
      expect(
        flattenMin({ xs: '0.5em', bp: '10em' }, width => ({ width }))
      ).toMatchSnapshot()
    })

    it('renders the renders width 0.5em without a media query', () => {
      expect(
        flattenMin('0.5em', width => ({ width }))
      ).toMatchSnapshot()
    })
  })

  describe('flattenMax', () => {
    it('renders the renders width 0.5em inside a max xl media query', () => {
      expect(
        flattenMax({ xl: '5em', bp: '10em' }, width => ({ width }))
      ).toMatchSnapshot()
    })

    it('renders the renders width 0.5em without a media query', () => {
      expect(
        flattenMax('0.5em', width => ({ width }))
      ).toMatchSnapshot()
    })
  })
})
