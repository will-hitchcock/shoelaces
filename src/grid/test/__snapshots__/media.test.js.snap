// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media between renders the styles inside a between md and lg media query 1`] = `
Object {
  "@media (min-width: 48em and (max-width: 62em": Object {
    "key": "value",
  },
}
`;

exports[`media between renders the styles without a media query 1`] = `
Object {
  "key": "value",
}
`;

exports[`media between renders the styles without a media query 2`] = `
Object {
  "key": "value",
}
`;

exports[`media between renders the styles without a media query 3`] = `
Object {
  "key": "value",
}
`;

exports[`media flattenMax renders the renders width 0.5em inside a max xl media query 1`] = `
Object {
  "@media (max-width: 75em)": Object {
    "width": "5em",
  },
}
`;

exports[`media flattenMax renders the renders width 0.5em without a media query 1`] = `
Object {
  "width": "0.5em",
}
`;

exports[`media flattenMin renders the renders width 0.5em inside a min xs media query 1`] = `
Object {
  "@media (min-width: 0)": Object {
    "width": "0.5em",
  },
}
`;

exports[`media flattenMin renders the renders width 0.5em without a media query 1`] = `
Object {
  "width": "0.5em",
}
`;

exports[`media max renders the styles inside a max md media query 1`] = `
Object {
  "@media (max-width: 48em)": Object {
    "key": "value",
  },
}
`;

exports[`media max renders the styles without a media query 1`] = `
Object {
  "key": "value",
}
`;

exports[`media min renders the styles inside a min md media query 1`] = `
Object {
  "@media (min-width: 48em)": Object {
    "key": "value",
  },
}
`;

exports[`media min renders the styles without a media query 1`] = `
Object {
  "key": "value",
}
`;

exports[`media min renders the styles without a media query 2`] = `
Object {
  "key": "value",
}
`;
